### DISCUSSION TITLE
I2C with pullups

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#11292] [Hardware & Peripherals]
I'm trying to find a small board with pull ups for the I2C lines. The wemos S2 pico has them but the display makes it a bigger board than I would like. The https://www.wemos.cc/en/latest/_static/files/sch_c3_mini_v2.1.0.pdf has a couple of pull ups for I02 & I08 but the I2C is on I010 & I08, similarly the https://www.wemos.cc/en/latest/_static/files/sch_c3_pico_v1.0.0.pdf. Maybe I could use a machine.SoftI2C on I02 & I08?

Does anybody know the purpose of apparently random 10k pull ups on these boards? The one on I018 in https://www.wemos.cc/en/latest/_static/files/sch_s2_mini_v1.0.0.pdf for example?

**Comment by peterhinch** [comment]
SoftI2C may be your answer. The reason for pullups on I/O pins is usually connected with the boot process, but you'd need to study the chip datasheet to be sure.

**Comment by robert-hh** [comment]
Did you try to use Pins 2 and 8 for I2C? The ESP32 devices usually allow free assignment of pins. At a ESP32C3 board here, the following statement is accepted:
`i=I2C(0, sda=Pin(2), scl=Pin(8))`
And a I2C device attached to it shows a valid address when scanned,

**Reply by kjm1102** [reply]
I've changed uart pins before Rob. Never tried it with I2C, I'll give it a try. With regard to I2C I've got an anonymous I2C water depth sensor with zero documentation. Instinct tells me I should try to lower the baud (given its on the end of a 5m cable) but other than seeing it drive displays I don't have a lot of I2C experience. Is there like a standard I2C chip that sensor manufacturers use that I could read up on?

**Reply by robert-hh** [reply]
I2C is not designed for long wires. the intended wire length is <= 15cm for on-PCB connections. You can do that, but you have to use the smallest feasible pull-up resistors and/or lower the board rate.Assuming a capacity of 500pF for a 5m cable and a 10k Pull-up resistor, the time constant is 50µs, the period of the I2c clock should be at least 200µs, to a suitable baud rate is about 5 kHz. That can easily done with SoftI2C, if required.
 And it will be subjective to distortion. That would be another reason to use as-low-as-possible pull-up resistors, like 470 Ohm. And use a shielded cable.

**Comment by dhalbert** [comment]
> Does anybody know the purpose of apparently random 10k pull ups on these boards? The one on I018 in https://www.wemos.cc/en/latest/_static/files/sch_s2_mini_v1.0.0.pdf for example?

GPIO18 is apparently used for serial bootloading as a TX line for UART1. It is a strong design suggestion by Espressifthat this pin be tied to 3.3V to prevent it floating. Their modules include that pullup in the module.
From https://www.espressif.com/sites/default/files/documentation/esp32-s2_hardware_design_guidelines_en.pdf:

> _GPIO18 works as U1RXD and is in an uncertain state when the chip is powered on, which may affect the chip’s entry into download boot mode. To solve this issue, add an external pull-up resistor._

That pullup also affects use of one of the DACs. We were just discussing that over here: https://github.com/adafruit/circuitpython/issues/7871

**Reply by kjm1102** [reply]
Tnx, that explains a lot.
